jupyter-extension/jupyterlab_pachyderm/pfs_manager.py

```diff
@@ -3,11 +3,84 @@
 from jupyter_server.services.contents.filemanager import FileContentsManager
 import mimetypes
 from pachyderm_sdk import Client
-from pachyderm_sdk.api import pfs
+from pachyderm_sdk.api import pfs, pps
+import os
 from pathlib import Path
 from tornado import web
 
 
+def _create_base_model(path: str, fileinfo: pfs.FileInfo, type: str) -> dict:
+    if type == "file" and fileinfo.file_type != pfs.FileType.FILE:
+        raise web.HTTPError(400, f"{path} is not a file", reason="bad type")
+    if type == "directory" and fileinfo.file_type != pfs.FileType.DIR:
+        raise web.HTTPError(400, f"{path} is not a directory", reason="bad type")
+
+    model = {}
+    model["name"] = Path(path).name
+    model["path"] = path
+    model["created"] = fileinfo.committed
+    model["last_modified"] = fileinfo.committed
+    model["type"] = "directory" if fileinfo.file_type == pfs.FileType.DIR else "file"
+    model["content"] = None
+    model["mimetype"] = None
+    model["format"] = None
+    model["writable"] = False
+
+    return model
+
+
+def _get_file_model(
+    client: Client, model: dict, path: str, file: pfs.File, format: str
+):
+    get_response = client.pfs.get_file(file=file)
```

Comment:
generally, yes the file will exist since we are obtaining the pfs.File either from getting the directory it exists in or as part of a list_datum response. there are edge cases where the file ceases to exist, e.g. if the mounted branch gets deleted or something. in those cases, i think it should be fine to bubble the exception up since there isn't really a clean way to handle those exceptions within the contents manager itself.